### PR TITLE
Added cmake build system 

### PR DESCRIPTION
**Pull request by FrostyMike**
- Added cmake-based build system for F_Front and F_Back
- Made c++ compatibility fixes to the F_front code

### CODE REVIEW

**Comment by h-murai** [comment]
Is it confirmed that the existing Makefile build works after this change is merged?

File: F-FrontEnd/gnu_module/trans_module.c [hunk 1]
```diff
@@ -766,7 +766,7 @@ static ID  make_id( gfc_symbol *sym , const char *name )
 
        if (sym->attr.flavor == FL_MODULE) {
    
-          id->class = CL_MODULE;
+          id->cls = CL_MODULE;
```
Comment (h-murai) [inline F-FrontEnd/gnu_module/trans_module.c]: Is this change for avoiding a C++ keyword "class"?

Comment (FrostyMike) [reply F-FrontEnd/gnu_module/trans_module.c]: yes

**Comment by FrostyMike** [comment]
> Is it confirmed that the existing Makefile build works after this change is merged?

It was tested successfully in travis using gcc-7 and gfortran-7.
https://travis-ci.org/github/omni-compiler/xcodeml-tools/builds/764388408#L3835

File: F-FrontEnd/src/F-ACC.c [hunk 2]
```diff
@@ -214,6 +214,42 @@ static enum ACC_pragma get_begin_directive(enum ACC_pragma dir)
...
+	case ACC_ENTER_DATA: return ACC_ENTER_DATA;
+	case ACC_EXIT_DATA: return ACC_EXIT_DATA;
+	case ACC_HOST_DATA: return ACC_HOST_DATA;
+	case ACC_END_HOST_DATA: return ACC_END_HOST_DATA;
+	case ACC_DECLARE: return ACC_DECLARE;
+	case ACC_UPDATE_D: return ACC_UPDATE_D;
+	case ACC_INIT: return ACC_INIT;
+	case ACC_SHUTDOWN: return ACC_SHUTDOWN;
+	case ACC_SET: return ACC_SET;
+    default:
```
Comment (clementval) [inline F-FrontEnd/src/F-ACC.c]: NIT: Indent looks odd here.

File: F-FrontEnd/gnu_module/export_module.c [hunk 3]
```diff
@@ -429,23 +429,21 @@ struct list_node *cons_list(x,l)
     return(lp);
 }
 
-expr list2(code,x1,x2)
-     enum expr_code code;
-     expr x1,x2;
+expr list2(enum expr_code code,
+	     expr x1,
```
Comment (clementval) [inline F-FrontEnd/gnu_module/export_module.c]: NIT: Is there any reason to have `expr x1` and `expr x2` on a separate line? Same comment for other places like that in this PR.

File: Driver/bin/F_Back.in [hunk 4]
```diff
@@ -18,14 +18,18 @@ function parseParameter()
...
+    OMNI_HOME="${SCRIPT_PATH}/.."
     OMNI_JAR1="${OMNI_HOME}/share/om-f-back.jar"
     OMNI_JAR2="${OMNI_HOME}/share/om-common.jar"
 else
     OMNI_JAR1="../F-BackEnd/build/om-f-back.jar"
     OMNI_JAR2="../XcodeML-Common/build/om-common.jar"
 fi
 
+if [ -z ${OMNI_JAVA+x} ]; then
 OMNI_JAVA=@JAVA@
```
Comment (clementval) [inline Driver/bin/F_Back.in]: NIT: Missing indent.

File: F-FrontEnd/src/F-XMP.c [hunk 5]
```diff
@@ -85,6 +85,43 @@ void init_for_XMP_pragma()
     XMP_io_desired_statements = 0;
 }
 
+static enum XMP_pragma to_XMP_pragma(omllint_t val)
+{
+	switch(val)
```
Comment (clementval) [inline F-FrontEnd/src/F-XMP.c]: NIT: Indent does not follow code in this file.

File: F-FrontEnd/gnu_module/export_module.c [hunk 6]
```diff
@@ -429,23 +429,21 @@ struct list_node *cons_list(x,l)
...
+		 expr x2)
+
 {
     return(make_enode(code,(void *)cons_list(x1,cons_list(x2,NULL))));
 }
 
-expr list3(code,x1,x2,x3)
-     enum expr_code code;
-     expr x1,x2,x3;
+expr list3(enum expr_code code, expr x1,expr x2,expr x3)
```
Comment (clementval) [inline F-FrontEnd/gnu_module/export_module.c]: NIT: Missing some whitespaces. 
```suggestion
expr list3(enum expr_code code, expr x1, expr x2, expr x3)
```

File: F-FrontEnd/src/C-ACC.h [hunk 7]
```diff
@@ -37,6 +37,8 @@ enum ACC_pragma {
     ACC_DIR_END
 };
 
+
```
Comment (clementval) [inline F-FrontEnd/src/C-ACC.h]: Not needed.

File: F-FrontEnd/src/F-intrinsics-table.c [hunk 8]
```diff
@@ -1617,7 +1617,7 @@ intrinsic_entry intrinsic_table[] = {
     {INTR_SELECTED_REAL_KIND,
      INTR_NAME_GENERIC,
      "selected_real_kind",
-     {INTR_TYPE_INT, INTR_TYPE_INT, INTR_INT},
+     {INTR_TYPE_INT, INTR_TYPE_INT, INTR_TYPE_INT},
```
Comment (clementval) [inline F-FrontEnd/src/F-intrinsics-table.c:1620]: This change seems to be unrelated to the cmake addition. Did you have error without changing this?

File: F-FrontEnd/src/hash.c [hunk 9]
```diff
@@ -103,12 +103,11 @@ static HashEntry *OneWordCreate(HashTable *tablePtr, const void *key,
  */
 
 void InitHashTable(
-    tablePtr,
-    keyType) register HashTable *tablePtr; /* Pointer to table record, which
-                                            * is supplied by the caller. */
-int keyType;                               /* Type of keys to use in table:
-                                            * HASH_STRING_KEYS, HASH_ONE_WORD_KEYS,
-                                            * or an integer >= 2. */
+        register HashTable *tablePtr, /* Pointer to table record, which
```
Comment (clementval) [inline F-FrontEnd/src/hash.c]: Indent and comment indent are messed up here.

File: F-FrontEnd/src/hash.c [hunk 10]
```diff
@@ -247,10 +245,9 @@ void DeleteHashTable(
  *----------------------------------------------------------------------
  */
 
-HashEntry *FirstHashEntry(tablePtr,
-                          searchPtr) HashTable *tablePtr; /* Table to search. */
-HashSearch *searchPtr; /* Place to store information about
-                        * progress through the table. */
+HashEntry *FirstHashEntry(HashTable *tablePtr, /* Table to search. */
+		HashSearch *searchPtr /* Place to store information about
+		                        * progress through the table. */)
```
Comment (clementval) [inline F-FrontEnd/src/hash.c]: NIT: Indent

File: F-FrontEnd/src/wrapper.cpp.in [hunk 11]
```diff
@@ -0,0 +1 @@
+#include "@SRC_FILE_BASENAME@.c"
```
Comment (clementval) [inline F-FrontEnd/src/wrapper.cpp.in]: NIT: Missing new line

File: F-FrontEnd/src/tests-runner.py [hunk 12]
```diff
@@ -0,0 +1,722 @@
+#!/usr/bin/env python3
+#-*- coding: utf-8 -*-
+
+__author__ = "Mikhail Zhigun"
+
+""" This module contains command-line test runner implementation for XcodeML tools, omni compiler Fortran frontend and 
+    backend. Should be run from unix shell with necessary command-line arguments. TestRunner can automatically scan 
+    for module dependencies and run tests in parallel. For debugging individual tests     run the tool with --input-test
```
Comment (clementval) [inline F-FrontEnd/src/tests-runner.py]: NIT: too many whitespaces

File: F-FrontEnd/src/tests-runner.py [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+     and --working-dir arguments. Then interim files from all test stages will be saved.
```
Comment (clementval) [inline F-FrontEnd/src/tests-runner.py]: NIT:  Line is not aligned with previous ones.

File: include/config.h.cmake.in [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+#cmakedefine HAVE_DOUBLE
+#ifdef HAVE_DOUBLE
+#cmakedefine SIZEOF_DOUBLE @SIZEOF_DOUBLE@
+#endif
+#cmakedefine HAVE_LONG_DOUBLE
+#ifdef HAVE_LONG_DOUBLE
+#cmakedefine SIZEOF_LONG_DOUBLE @SIZEOF_LONG_DOUBLE@
+#endif
+
+#endif //OMNI_CONFIG_H
```
Comment (clementval) [inline include/config.h.cmake.in]: NIT: Missing new line

**Review by clementval — COMMENTED** [review]
Couple of NITs about the indent and whitespaces in modified code. 

The title of the PR is a bit misleading since it modifies lots of C files and not just add cmake build support. Might be retitled with a clearer message or just in the commits. 

@h-murai I'll let you give the final approval.

File: include/config.h.cmake.in (hunk 14 above)
Comment (FrostyMike) [reply include/config.h.cmake.in]: Fixed

File: F-FrontEnd/src/wrapper.cpp.in (hunk 11 above)
Comment (FrostyMike) [reply F-FrontEnd/src/wrapper.cpp.in]: Fixed

File: F-FrontEnd/src/F-intrinsics-table.c (hunk 8 above)
Comment (FrostyMike) [reply F-FrontEnd/src/F-intrinsics-table.c:1620]: This is part of "Made c++ compatibility fixes to the F_front code" (from description of this PR). There was an error here, as "INTR_INT" is from entirely different enum. I guess it went unreported, because  C compilers are more lax about type safety.

**Comment by h-murai** [comment]
@FrostyMike Please follow the comments before I approve this PR.

File: F-FrontEnd/src/F-ACC.c (hunk 2 above)
Comment (FrostyMike) [reply F-FrontEnd/src/F-ACC.c]: Fixed (formatted it same way as previous switch)

File: F-FrontEnd/gnu_module/export_module.c (hunk 3 above)
Comment (FrostyMike) [reply F-FrontEnd/gnu_module/export_module.c]: Reverted changes to this file

File: Driver/bin/F_Back.in (hunk 4 above)
Comment (FrostyMike) [reply Driver/bin/F_Back.in]: Fixed

File: F-FrontEnd/gnu_module/export_module.c (hunk 6 above)
Comment (FrostyMike) [reply F-FrontEnd/gnu_module/export_module.c]: Reverted changes to this file

File: F-FrontEnd/src/tests-runner.py (hunk 13 above)
Comment (FrostyMike) [reply F-FrontEnd/src/tests-runner.py]: Fixed

File: F-FrontEnd/src/tests-runner.py (hunk 12 above)
Comment (FrostyMike) [reply F-FrontEnd/src/tests-runner.py]: Fixed

File: F-FrontEnd/src/hash.c (hunk 10 above)
Comment (FrostyMike) [reply F-FrontEnd/src/hash.c]: Fixed (as in "made it look similar to what was before")

File: F-FrontEnd/src/C-ACC.h (hunk 7 above)
Comment (FrostyMike) [reply F-FrontEnd/src/C-ACC.h]: Fixed

File: F-FrontEnd/src/hash.c (hunk 9 above)
Comment (FrostyMike) [reply F-FrontEnd/src/hash.c]: Fixed (max similar to what was before)

File: F-FrontEnd/src/F-XMP.c (hunk 5 above)
Comment (FrostyMike) [reply F-FrontEnd/src/F-XMP.c]: Fixed (I assume the problem was tabs instead of spaces)

**Comment by FrostyMike** [comment]
> @FrostyMike Please follow the comments before I approve this PR.

Done

**Comment by h-murai** [comment]
@FrostyMike I merged this PR once but then reverted it. It's because I noticed that this new cmake build works for F_Front and F_Back but doesn't for C_Front and C_Back. The build system on the top level directory must be able to handle all of them, or, as the second best, you have to note the limitation in a document such as README. What do you think about this?

**Comment by FrostyMike** [comment]
> @FrostyMike I merged this PR once but then reverted it. It's because I noticed that this new cmake build works for F_Front and F_Back but doesn't for C_Front and C_Back. The build system on the top level directory must be able to handle all of them, or, as the second best, you have to note the limitation in a document such as README. What do you think about this?

@h-murai Fine, I've added CMake based build for C_Front and C_Back too, see https://github.com/omni-compiler/xcodeml-tools/pull/177

Please note however, that this is outside the scope of our agreement, as it was previously discussed that I will work only on F_Front. As MeteoSwiss doesn't use C_Front, I am not really authorised to spend a lot of development time on it. So if you require further changes, please discuss it with my supervisor @lxavier

**Comment by h-murai** [comment]
@FrostyMike Thank you very much for your additional efforts. I understand that C_Front and C_Back are out of your target, but we have to avoid confusing users about how to build xcodeml-tools. I'll be responsible for maintaining the build system for C_Front and C_Back that you kindly added.
